x/username/handlers.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	token := Token{
+		Metadata: &weave.Metadata{Schema: 1},
+		Targets:  msg.Targets,
+		Owner:    owner,
+	}
+	if _, err := h.bucket.Put(db, msg.Username.Bytes(), &token); err != nil {
+		return nil, errors.Wrap(err, "cannot store token")
+	}
+	return &weave.DeliverResult{Data: msg.Username.Bytes()}, nil
+}
+
+func (h *registerTokenHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*RegisterTokenMsg, error) {
+	var msg RegisterTokenMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	switch err := h.bucket.Has(db, msg.Username.Bytes()); {
+	case err == nil:
+		return nil, errors.Wrapf(errors.ErrDuplicate, "username %q already registered", msg.Username)
+	case errors.ErrNotFound.Is(err):
+		// All good. Username is not taken yet.
+	default:
+		return nil, errors.Wrap(err, "cannot check if username is unique")
+	}
+	return &msg, nil
+}
+
+type changeTokenOwnerHandler struct {
+	auth   x.Authenticator
+	bucket orm.ModelBucket
+}
+
+func (h *changeTokenOwnerHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: changeTokenOwnerCost}, nil
+}
+
+func (h *changeTokenOwnerHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, token, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	token.Owner = msg.NewOwner
+	if _, err := h.bucket.Put(db, msg.Username.Bytes(), token); err != nil {
+		return nil, errors.Wrap(err, "cannot store token")
+	}
+	return &weave.DeliverResult{Data: msg.Username.Bytes()}, nil
+}
+
+func (h *changeTokenOwnerHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*ChangeTokenOwnerMsg, *Token, error) {
+	var msg ChangeTokenOwnerMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, nil, errors.Wrap(err, "load msg")
+	}
+
+	var token Token
+	if err := h.bucket.One(db, msg.Username.Bytes(), &token); err != nil {
+		return nil, nil, errors.Wrap(err, "cannot get token from database")
+	}
+
+	if !h.auth.HasAddress(ctx, token.Owner) {
+		return nil, nil, errors.Wrap(errors.ErrUnauthorized, "only the token owner can execute this operation")
+	}
+
+	return &msg, &token, nil
+}
+
+type changeTokenTargetsHandler struct {
+	auth   x.Authenticator
+	bucket orm.ModelBucket
+}
+
+func (h *changeTokenTargetsHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: changeTokenTargetCost}, nil
+}
+
+func (h *changeTokenTargetsHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, token, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	token.Targets = msg.NewTargets
```

Comment:
The code looks good and addresses all my concerns.
The one issue I have is here. And this is a design question.

As it is currently written in (old) nft/username, it is a PATCH to add or remove names.
This defines it as a POST to overwrite them.

POST is much simpler to do on the server. But makes it a bit harder for the client to perform the action "add an ethereum address to my username".

Reply:
We can keep it as you implemented it. Just wanted to bring it up as a discussion point

Reply:
This is a great question. I always tend to do overwrite (POST) for two reasons.
1. It is a tiny bit easier to implement and much easier to explain how it works,
2. The client is always having a full state. Sending the full list should never be an issue. Sometimes it might be even easier than building the diff.

If this is an issue, we can always change it or introduce another message which provides different update strategy.

Reply:
We should not push complexity to the client. A full overwrite is problematic in any WYSIWYG environment or where state gets updated quickly. I also want to highlight the asynchronous nature of our system. All the TXs get queues before they are applied and the order is not static as well.
A patch would make the system more resilient as we would not overwrite by adds and could fail on non existing removes.

Reply:
Yeah, I think I'm guilty of suggesting the full overwrite too as it seemed easier to reason about in terms of server-side implementation, but indeed in blockchain case it might be wiser to do a patch.

Reply:
Who is the person to decide how this should work? @kmw101 ?

Reply:
I have discussed this with @alpe and the `PATCH` solution would require the following rules

```
// ChangeTokenTargetsMsg is a request to change the address that this token
// points to. Only the owner of a token can request this operation.
message ChangeUsernameTokenTargetsMsg {
  weave.Metadata metadata = 1;
  // Username is the unique name of the token, for example alice*iov
  string username = 2 [(gogoproto.casttype) = "Username"];
  //
  // - if the .Address  is not empty then set/overwrite. 
  // - if the .Address is empty (nil) then delete 
  // - .Blokchain attribute value must be unique in the list.
  // - the new .Address value can be the same as the current value
  // - deleting an item that does not exist fails
  
  // In a different message (register)
  // - you cannot register a token with a list containing an empty .Blockchain value
  repeated BlockchainAddress diff = 3 [(gogoproto.nullable) = false];
}
```
The problem with this approach is that it is a very complex logic to implement. The current regexp + lenght check must be replaced with custom set of rules, different for each message.
The biggest issues are:

1. [`BlockchainAddress.Validate`](https://github.com/iov-one/weave/blob/e8238a3da70c0614b162665a962705b6c5dbb139/cmd/bnsd/x/username/model.go#L17-L27) can no longer validate the state as the validation rules differs depending on the message that is using it,
2. You need to have access to the database in order validate the `BlockchainAddress` collection,
3. It is easy to create an invalid configuration, not only by using an invalid data (ie invalid blockchain ID), but by providing a setup that leads to a logic violation

This is the last outstanding issue in this pull request. Should we proceed with the above implementation, try something else or merge the current state?

Reply:
Post vs Patch is not blocking this PR IMHO. It would be nice to have it for the client but we should stick to our timeline. Patch can to into a new PR if we want to do it.

Reply:
Unless there will be any objections, I will merge this within an hour. :racehorse: